finetuner-workflow/finetuner/finetuner.py

```diff
@@ -1054,6 +1076,7 @@ def collector(data):
 # At the end of it all, record to a `final` output.
 final_path = os.path.join(output_dir, "final")
 trainer.save_model(final_path)
+tensorizer_save(final_path) # Must be invoked manually.
```

Comment:
If `save_steps` is a multiple of the number of steps per epoch, unless I'm unaware of how `Trainer` saves checkpoints, it will do a double save at the end, or two quick saves at the end if `num_steps % num_steps_per_epoch` is close to 0. Do we want to possibly skip saving the final checkpoint if this is the case?

Reply:
For predictability, there are pros to not skipping it; then something exists at the checkpoint file path and at the `final_path` path, and either can be deleted safely, or otherwise processed programmatically. Though in a storage location where hardlinking is available, that could be used to reduce data duplication while keeping both paths (if it is safe to propagate modifications to either file to either other file).

Reply:
The checkpoint callback is not called if `trainer.save_model()` is invoked manually.